riscv/imsic.cc

```diff
@@ -0,0 +1,177 @@
+      csrmap[IMSIC_EIE(i * 2)] = std::make_shared<rv32_low_csr_t>(proc, IMSIC_EIE(i * 2), eie[i]);
+      csrmap[IMSIC_EIE(i * 2 + 1)] = std::make_shared<rv32_high_csr_t>(proc, IMSIC_EIE(i * 2 + 1), eie[i]);
+    } else {
+      csrmap[IMSIC_EIP(i * 2)] = eip[i];
+      csrmap[IMSIC_EIE(i * 2)] = eie[i];
+    }
+  }
+}
+
+reg_t imsic_file_t::topei() {
+  reg_t thd = eithreshold->read();
+  for (size_t i = 0; i < IMSIC_NUM_EI_REGS; i++) {
+    reg_t ints = eip[i]->read() & eie[i]->read();
+    if (ints == 0)
+      continue;
+    reg_t iid = ctz(ints) + i * 64;
+    // When eithreshold is a nonzero value P, interrupt identities P and higher do not contribute to signaling interrupts
+    if (thd && iid >= thd)
+      return 0;
+    return iid;
+  }
+  return 0;
+}
+
+void imsic_file_t::claimei(reg_t intr) {
+  if (intr > 0 && intr < 64 * IMSIC_NUM_EI_REGS) {
+    size_t reg = intr / 64;
+    size_t idx = intr % 64;
+    eip[reg]->write(eip[reg]->read() & ~(reg_t(1) << idx));
+  }
+}
+
+void imsic_file_t::pendei(reg_t intr) {
+  if (intr > 0 && intr < 64 * IMSIC_NUM_EI_REGS) {
+    size_t reg = intr / 64;
+    size_t idx = intr % 64;
+    eip[reg]->write(eip[reg]->read() | (reg_t(1) << idx));
+  }
+}
+
+void imsic_file_t::update_mip() {
+  reg_t iid = eidelivery->read() ? topei() : 0;
+  if (v) {
+    // Privileged 9.2.4: Register hgeie selects the subset of guest external interrupts that cause a supervisor-level (HS-level) guest external interrupt.
+    bool sgeip = proc->get_state()->hgeie->read() & proc->get_state()->hgeip->read();
+    state->mip->backdoor_write_with_mask(MIP_SGEIP, sgeip ? MIP_SGEIP : 0);
+  }
+  // The enable bits in hgeie do not affect the VS-level external interrupt signal selected from hgeip by hstatus.VGEIN
+  if (!v || (v && vgein == get_field(state->hstatus->read(), HSTATUS_VGEIN))) {
+    // if not virtual, or virtual and vgein is selected in hstatus
+    state->mip->backdoor_write_with_mask(mip_mask, iid ? mip_mask : 0);
+  }
+}
+
+imsic_mmio_t::imsic_mmio_t(imsic_file_t_p const imsic) : imsic(imsic) {
+}
+
+bool imsic_mmio_t::load(reg_t addr UNUSED, size_t len, uint8_t* bytes) {
+  if (len != 4)
+    return false;
+  // return 0 in all cases
+  memset(bytes, 0, 4);
+  return true;
+}
+
+bool imsic_mmio_t::store(reg_t addr, size_t len, const uint8_t* bytes) {
+  if (len != 4)
+    return false;
+  if (addr == SETEIPNUM_LE || addr == SETEIPNUM_BE) {
+    uint32_t intr;
+    memcpy(&intr, bytes, sizeof(intr));
+    if (addr == SETEIPNUM_BE)
+      intr = swap(intr);
+    imsic->pendei(intr);
+  }
+  return true;
+}
+
+imsic_t::imsic_t(processor_t *proc, unsigned geilen) {
+  if (proc->extension_enabled_const(EXT_SMAIA))
+    m = std::make_shared<imsic_file_t>(proc, MIP_MEIP, IMSIC_M_FILE_REGS);
+  if (proc->extension_enabled_const(EXT_SSAIA)) {
+    s = std::make_shared<imsic_file_t>(proc, MIP_SEIP, IMSIC_S_FILE_REGS);
+    assert(geilen <= 63);
```

Comment:
[nitpick] This would be much more readable:
```suggestion
    assert(geilen < 64);
```